File: push/src/main/java/com/xiaomi/mipush/sdk/PushMessageHandler.java
package com.xiaomi.mipush.sdk;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ResolveInfo;
import android.os.IBinder;
import android.text.TextUtils;
import com.xiaomi.channel.commonutils.android.SystemUtils;
import com.xiaomi.channel.commonutils.logger.MyLog;
import com.xiaomi.mipush.sdk.MessageHandleService;
import com.xiaomi.mipush.sdk.MiPushClient;
import com.xiaomi.push.service.PushConstants;
import com.xiaomi.push.service.PushServiceConstants;
import com.xiaomi.push.service.clientReport.PushClientReportManager;
import com.xiaomi.push.service.xmpush.Command;
import com.xiaomi.xmpush.thrift.ClientUploadDataItem;
import com.xiaomi.xmpush.thrift.XmPushThriftSerializeUtils;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

/* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/mipush/sdk/PushMessageHandler.class */
public class PushMessageHandler extends BaseService {
    private static List<MiPushClient.ICallbackResult<?>> sICallbackResult = new ArrayList<>();
    private static List<MiPushClient.MiPushClientCallback> sCallbacks = new ArrayList<>();
    private static ThreadPoolExecutor sPool = new ThreadPoolExecutor(1, 1, 15, TimeUnit.SECONDS, new LinkedBlockingQueue<>());

    /* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/mipush/sdk/PushMessageHandler$PushMessageInterface.class */
    interface PushMessageInterface extends Serializable {
    }

    public static void addJob(Context context, Intent intent) {
        MyLog.v("addjob PushMessageHandler " + intent);
        if (intent != null) {
            scheduleJob(context, intent);
            startService(context);
        }
    }

    protected static void addPushCallbackClass(MiPushClient.MiPushClientCallback miPushClientCallback) {
        synchronized (sCallbacks) {
            if (!sCallbacks.contains(miPushClientCallback)) {
                sCallbacks.add(miPushClientCallback);
            }
        }
    }

    protected static void addUPSCallback(MiPushClient.ICallbackResult<?> iCallbackResult) {
        synchronized (sICallbackResult) {
            if (!sICallbackResult.contains(iCallbackResult)) {
                sICallbackResult.add(iCallbackResult);
            }
        }
    }

    private static void handleNewMessage(Context context, Intent intent, ResolveInfo resolveInfo) {
        try {
            MessageHandleService.addJob(context.getApplicationContext(), new MessageHandleService.MessageHandleJob(intent, (PushMessageReceiver) SystemUtils.loadClass(context, resolveInfo.activityInfo.name).getDeclaredConstructor().newInstance()));
            MessageHandleService.onHandleIntent(context, new Intent(context.getApplicationContext(), (Class<?>) MessageHandleService.class));
        } catch (Throwable th) {
            MyLog.e(th);
        }
    }

    public static boolean isCallbackEmpty() {
        return sCallbacks.isEmpty();
    }

    protected static boolean isCategoryMatch(String str, String str2) {
        return (TextUtils.isEmpty(str) && TextUtils.isEmpty(str2)) || TextUtils.equals(str, str2);
    }

    protected static void onCommandResult(Context context, String str, String str2, long j, String str3, List<String> list) {
        synchronized (sCallbacks) {
            for (MiPushClient.MiPushClientCallback miPushClientCallback : sCallbacks) {
                if (isCategoryMatch(str, miPushClientCallback.getCategory())) {
                    miPushClientCallback.onCommandResult(str2, j, str3, list);
                }
            }
        }
    }

    protected static void onHandleIntent(Context context, Intent intent) {
        try {
            if (PushConstants.ACTION_WAKEUP.equals(intent.getAction())) {
                AwakeHelper.doAWork(context, intent, null);
            } else if (PushConstants.MIPUSH_ACTION_SEND_TINYDATA.equals(intent.getAction())) {
                ClientUploadDataItem clientUploadDataItem = new ClientUploadDataItem();
                XmPushThriftSerializeUtils.convertByteArrayToThriftObject(clientUploadDataItem, intent.getByteArrayExtra(PushConstants.MIPUSH_EXTRA_PAYLOAD));
                MyLog.v("PushMessageHandler.onHandleIntent " + clientUploadDataItem.getId());
                MiTinyDataClient.upload(context, clientUploadDataItem);
            } else if (1 == PushMessageHelper.getPushMode(context)) {
                if (isCallbackEmpty()) {
                    MyLog.e("receive a message before application calling initialize");
                } else {
                    PushMessageInterface pushMessageInterfaceProcessIntent = PushMessageProcessor.getInstance(context).processIntent(intent);
                    if (pushMessageInterfaceProcessIntent != null) {
                        processMessageForCallback(context, pushMessageInterfaceProcessIntent);
                    }
                }
            } else if (!PushServiceConstants.ACTION_SYNC_LOG.equals(intent.getAction())) {
                Intent intent2 = new Intent(PushConstants.MIPUSH_ACTION_NEW_MESSAGE);
                intent2.setPackage(context.getPackageName());
                intent2.putExtras(intent);
                try {
                    List<ResolveInfo> listQueryBroadcastReceivers = context.getPackageManager().queryBroadcastReceivers(intent2, 32);
                    ResolveInfo next = null;
                    if (listQueryBroadcastReceivers != null) {
                        Iterator<ResolveInfo> it = listQueryBroadcastReceivers.iterator();
                        while (true) {
                            next = null;
                            if (!it.hasNext()) {
                                break;
                            }
                            next = it.next();
                            if (next.activityInfo != null && next.activityInfo.packageName.equals(context.getPackageName()) && PushMessageReceiver.class.isAssignableFrom(SystemUtils.loadClass(context, next.activityInfo.name))) {
                                break;
                            }
                        }
                    }
                    if (next != null) {
                        handleNewMessage(context, intent2, next);
                    } else {
                        MyLog.e("cannot find the receiver to handler this message, check your manifest");
                        PushClientReportManager.getInstance(context).reportEvent4ERROR(context.getPackageName(), intent, "11");
                    }
                } catch (Exception e) {
                    MyLog.e(e);
                    PushClientReportManager.getInstance(context).reportEvent4ERROR(context.getPackageName(), intent, "9");
                }
            }
        } catch (Throwable th) {
            MyLog.e(th);
            PushClientReportManager.getInstance(context).reportEvent4ERROR(context.getPackageName(), intent, "10");
        }
    }

    public static void onInitializeResult(long j, String str, String str2) {
        synchronized (sCallbacks) {
            Iterator<MiPushClient.MiPushClientCallback> it = sCallbacks.iterator();
            while (it.hasNext()) {
                it.next().onInitializeResult(j, str, str2);
            }
        }
    }

    public static void onReceiveMessage(Context context, MiPushMessage miPushMessage) {
        synchronized (sCallbacks) {
            for (MiPushClient.MiPushClientCallback miPushClientCallback : sCallbacks) {
                if (isCategoryMatch(miPushMessage.getCategory(), miPushClientCallback.getCategory())) {
                    miPushClientCallback.onReceiveMessage(miPushMessage.getContent(), miPushMessage.getAlias(), miPushMessage.getTopic(), miPushMessage.isNotified());
                    miPushClientCallback.onReceiveMessage(miPushMessage);
                }
            }
        }
    }

    protected static void onSubscribeResult(Context context, String str, long j, String str2, String str3) {
        synchronized (sCallbacks) {
            for (MiPushClient.MiPushClientCallback miPushClientCallback : sCallbacks) {
                if (isCategoryMatch(str, miPushClientCallback.getCategory())) {
                    miPushClientCallback.onSubscribeResult(j, str2, str3);
                }
            }
        }
    }

    protected static void onUPSRegisterResult(Context context, MiPushCommandMessage miPushCommandMessage) {
        synchronized (sICallbackResult) {
            for (MiPushClient.ICallbackResult<?> iCallbackResult : sICallbackResult) {
                if (iCallbackResult instanceof MiPushClient.UPSRegisterCallBack) {
                    MiPushClient.UPSRegisterCallBack registerCallBack = (MiPushClient.UPSRegisterCallBack) iCallbackResult;
                    MiPushClient.TokenResult tokenResult = new MiPushClient.TokenResult();
                    if (miPushCommandMessage != null && miPushCommandMessage.getCommandArguments() != null && miPushCommandMessage.getCommandArguments().size() > 0) {
                        tokenResult.setResultCode(miPushCommandMessage.getResultCode());
                        tokenResult.setToken(miPushCommandMessage.getCommandArguments().get(0));
                    }
                    registerCallBack.onResult(tokenResult);
                }
            }
        }
    }

    protected static void onUnsubscribeResult(Context context, String str, long j, String str2, String str3) {
        synchronized (sCallbacks) {
            for (MiPushClient.MiPushClientCallback miPushClientCallback : sCallbacks) {
                if (isCategoryMatch(str, miPushClientCallback.getCategory())) {
                    miPushClientCallback.onUnsubscribeResult(j, str2, str3);
                }
            }
        }
    }

    public static void processMessageForCallback(Context context, PushMessageInterface pushMessageInterface) {
        if (pushMessageInterface instanceof MiPushMessage) {
            onReceiveMessage(context, (MiPushMessage) pushMessageInterface);
            return;
        }
        if (pushMessageInterface instanceof MiPushCommandMessage) {
            MiPushCommandMessage miPushCommandMessage = (MiPushCommandMessage) pushMessageInterface;
            String command = miPushCommandMessage.getCommand();
            if (Command.COMMAND_REGISTER.value.equals(command)) {
                List<String> commandArguments = miPushCommandMessage.getCommandArguments();
                String str = null;
                if (commandArguments != null) {
                    str = null;
                    if (!commandArguments.isEmpty()) {
                        str = commandArguments.get(0);
                    }
                }
                onInitializeResult(miPushCommandMessage.getResultCode(), miPushCommandMessage.getReason(), str);
                return;
            }
            if (Command.COMMAND_SET_ALIAS.value.equals(command) || Command.COMMAND_UNSET_ALIAS.value.equals(command) || Command.COMMAND_SET_ACCEPT_TIME.value.equals(command)) {
                onCommandResult(context, miPushCommandMessage.getCategory(), command, miPushCommandMessage.getResultCode(), miPushCommandMessage.getReason(), miPushCommandMessage.getCommandArguments());
                return;
            }
            if (Command.COMMAND_SUBSCRIBE_TOPIC.value.equals(command)) {
                List<String> commandArguments2 = miPushCommandMessage.getCommandArguments();
                onSubscribeResult(context, miPushCommandMessage.getCategory(), miPushCommandMessage.getResultCode(), miPushCommandMessage.getReason(), (commandArguments2 == null || commandArguments2.isEmpty()) ? null : commandArguments2.get(0));
            } else if (Command.COMMAND_UNSUBSCRIBE_TOPIC.value.equals(command)) {
                List<String> commandArguments3 = miPushCommandMessage.getCommandArguments();
                onUnsubscribeResult(context, miPushCommandMessage.getCategory(), miPushCommandMessage.getResultCode(), miPushCommandMessage.getReason(), (commandArguments3 == null || commandArguments3.isEmpty()) ? null : commandArguments3.get(0));
            }
        }
    }

    protected static void removeAllPushCallbackClass() {
        synchronized (sCallbacks) {
            sCallbacks.clear();
        }
    }

    protected static void removeAllUPSCallback() {
        synchronized (sICallbackResult) {
            sICallbackResult.clear();
        }
    }

    protected static void removePushCallbackClass(MiPushClient.MiPushClientCallback miPushClientCallback) {
        synchronized (sCallbacks) {
            sCallbacks.remove(miPushClientCallback);
        }
    }

    protected static void removeUPSCallback(MiPushClient.ICallbackResult<?> iCallbackResult) {
        synchronized (sICallbackResult) {
            sICallbackResult.remove(iCallbackResult);
        }
    }

    private static void scheduleJob(final Context context, final Intent intent) {
        if (intent == null || sPool.isShutdown()) {
            return;
        }
        sPool.execute(new Runnable() { // from class: com.xiaomi.mipush.sdk.PushMessageHandler.1
            @Override // java.lang.Runnable
            public void run() {
                PushMessageHandler.onHandleIntent(context, intent);
            }
        });
    }

    public static void startService(Context context) {
        Intent intent = new Intent();
        intent.setComponent(new ComponentName(context, (Class<?>) PushMessageHandler.class));
        try {
            context.startService(intent);
        } catch (Exception e) {
            MyLog.w(e.getMessage());
        }
    }

    @Override // com.xiaomi.mipush.sdk.BaseService
    protected boolean hasJob() {
        ThreadPoolExecutor threadPoolExecutor = sPool;
        return (threadPoolExecutor == null || threadPoolExecutor.getQueue() == null || sPool.getQueue().size() <= 0) ? false : true;
    }

    @Override // com.xiaomi.mipush.sdk.BaseService, android.app.Service
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override // com.xiaomi.mipush.sdk.BaseService, android.app.Service
    public int onStartCommand(Intent intent, int flags, int startId) {
        int result = super.onStartCommand(intent, flags, startId);
        scheduleJob(getApplicationContext(), intent);
        return result;
    }
}

File: legacy/src/main/java/com/xiaomi/mipush/sdk/PushMessageHandler.kt
package com.xiaomi.mipush.sdk

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.pm.ResolveInfo
import android.os.IBinder
import android.text.TextUtils
import com.xiaomi.channel.commonutils.android.SystemUtils
import com.xiaomi.channel.commonutils.logger.MyLog
import com.xiaomi.push.service.PushConstants
import com.xiaomi.push.service.PushServiceConstants
import com.xiaomi.push.service.clientReport.PushClientReportManager
import com.xiaomi.push.service.xmpush.Command
import com.xiaomi.xmpush.thrift.ClientUploadDataItem
import com.xiaomi.xmpush.thrift.XmPushThriftSerializeUtils
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

/*
 * Stock reference: com.xiaomi.xmsf 7.4.67-C (versionCode 70004067),
 * split-XiaomiServiceFrameworkCN-master.apk sha256 444e9f128591e04e38672bfe44a246ab3fa97ae68e95882839d8a7afe766df2b,
 * JADX path: com.xiaomi.xmsf/stock/split-XiaomiServiceFrameworkCN-master/sources/com/xiaomi/mipush/sdk/PushMessageHandler.java
 * Current override same-path: com.xiaomi.xmsf/current/base/sources/com/xiaomi/mipush/sdk/PushMessageHandler.java
 */
class PushMessageHandler : BaseService() {
    companion object {
        private val sICallbackResult = ArrayList<MiPushClient.ICallbackResult<*>>()
        private val sCallbacks = ArrayList<MiPushClient.MiPushClientCallback>()
        private val sPool = ThreadPoolExecutor(1, 1, 15, TimeUnit.SECONDS, LinkedBlockingQueue<Runnable>())

        @JvmStatic
        fun addJob(context: Context, intent: Intent) {
            MyLog.v("addjob PushMessageHandler $intent")
            scheduleJob(context, intent)
            startService(context)
        }

        @JvmStatic
        fun addPushCallbackClass(miPushClientCallback: MiPushClient.MiPushClientCallback) {
            synchronized(sCallbacks) {
                if (!sCallbacks.contains(miPushClientCallback)) {
                    sCallbacks.add(miPushClientCallback)
                }
            }
        }

        @JvmStatic
        fun addUPSCallback(iCallbackResult: MiPushClient.ICallbackResult<*>) {
            synchronized(sICallbackResult) {
                if (!sICallbackResult.contains(iCallbackResult)) {
                    sICallbackResult.add(iCallbackResult)
                }
            }
        }

        private fun handleNewMessage(context: Context, intent: Intent, resolveInfo: ResolveInfo) {
            try {
                MessageHandleService.addJob(
                    context.applicationContext,
                    MessageHandleService.MessageHandleJob(
                        intent,
                        SystemUtils.loadClass(context, resolveInfo.activityInfo.name).getDeclaredConstructor().newInstance() as PushMessageReceiver
                    )
                )
                MessageHandleService.onHandleIntent(context, Intent(context.applicationContext, MessageHandleService::class.java))
            } catch (th: Throwable) {
                MyLog.e(th)
            }
        }

        @JvmStatic
        fun isCallbackEmpty(): Boolean = sCallbacks.isEmpty()

        @JvmStatic
        fun isCategoryMatch(str: String?, str2: String?): Boolean {
            return (TextUtils.isEmpty(str) && TextUtils.isEmpty(str2)) || TextUtils.equals(str, str2)
        }

        @JvmStatic
        fun onCommandResult(context: Context, category: String, command: String, resultCode: Long, reason: String?, list: List<String>?) {
            synchronized(sCallbacks) {
                for (callback in sCallbacks) {
                    if (isCategoryMatch(category, callback.category)) {
                        callback.onCommandResult(command, resultCode, reason, list)
                    }
                }
            }
        }

        @JvmStatic
        fun onHandleIntent(context: Context, intent: Intent) {
            try {
                if (PushConstants.ACTION_WAKEUP == intent.action) {
                    AwakeHelper.doAWork(context, intent, null)
                } else if (PushConstants.MIPUSH_ACTION_SEND_TINYDATA == intent.action) {
                    val clientUploadDataItem = ClientUploadDataItem()
                    XmPushThriftSerializeUtils.convertByteArrayToThriftObject(
                        clientUploadDataItem,
                        intent.getByteArrayExtra(PushConstants.MIPUSH_EXTRA_PAYLOAD)
                    )
                    MyLog.v("PushMessageHandler.onHandleIntent ${clientUploadDataItem.id}")
                    MiTinyDataClient.upload(context, clientUploadDataItem)
                } else if (PushMessageHelper.getPushMode(context) == 1) {
                    if (isCallbackEmpty()) {
                        MyLog.e("receive a message before application calling initialize")
                    } else {
                        val pushMessageInterface = com.xiaomi.push.service.XMPushService.observer?.processMIPushIntent(intent) as? PushMessageInterface
                        if (pushMessageInterface != null) {
                            processMessageForCallback(context, pushMessageInterface)
                        }
                    }
                } else if (PushServiceConstants.ACTION_SYNC_LOG != intent.action) {
                    val intent2 = Intent(PushConstants.MIPUSH_ACTION_NEW_MESSAGE).apply {
                        setPackage(context.packageName)
                        putExtras(intent)
                    }
                    try {
                        val list = context.packageManager.queryBroadcastReceivers(intent2, 32)
                        var next: ResolveInfo? = null
                        for (item in list) {
                            if (item.activityInfo != null && item.activityInfo.packageName == context.packageName &&
                                PushMessageReceiver::class.java.isAssignableFrom(SystemUtils.loadClass(context, item.activityInfo.name))
                            ) {
                                next = item
                                break
                            }
                        }
                        if (next != null) {
                            handleNewMessage(context, intent2, next)
                        } else {
                            MyLog.e("cannot find the receiver to handler this message, check your manifest")
                            PushClientReportManager.getInstance(context).reportEvent4ERROR(context.packageName, intent, "11")
                        }
                    } catch (e: Exception) {
                        MyLog.e(e)
                        PushClientReportManager.getInstance(context).reportEvent4ERROR(context.packageName, intent, "9")
                    }
                }
            } catch (th: Throwable) {
                MyLog.e(th)
                PushClientReportManager.getInstance(context).reportEvent4ERROR(context.packageName, intent, "10")
            }
        }

        @JvmStatic
        fun onInitializeResult(j: Long, str: String?, str2: String?) {
            synchronized(sCallbacks) {
                for (callback in sCallbacks) {
                    callback.onInitializeResult(j, str, str2)
                }
            }
        }

        @JvmStatic
        fun onReceiveMessage(context: Context, miPushMessage: MiPushMessage) {
            synchronized(sCallbacks) {
                for (callback in sCallbacks) {
                    if (isCategoryMatch(miPushMessage.category, callback.category)) {
                        callback.onReceiveMessage(miPushMessage.content, miPushMessage.alias, miPushMessage.topic, miPushMessage.isNotified)
                        callback.onReceiveMessage(miPushMessage)
                    }
                }
            }
        }

        @JvmStatic
        fun onSubscribeResult(context: Context, category: String, resultCode: Long, reason: String?, topic: String?) {
            synchronized(sCallbacks) {
                for (callback in sCallbacks) {
                    if (isCategoryMatch(category, callback.category)) {
                        callback.onSubscribeResult(resultCode, reason, topic)
                    }
                }
            }
        }

        @JvmStatic
        fun onUPSRegisterResult(context: Context, miPushCommandMessage: MiPushCommandMessage?) {
            synchronized(sICallbackResult) {
                for (callback in sICallbackResult) {
                    if (callback is MiPushClient.UPSRegisterCallBack) {
                        val tokenResult = MiPushClient.TokenResult()
                        miPushCommandMessage?.let {
                            val args = it.commandArguments
                            if (!args.isNullOrEmpty()) {
                                tokenResult.setResultCode(it.resultCode)
                                tokenResult.setToken(args[0])
                            }
                        }
                        callback.onResult(tokenResult)
                    }
                }
            }
        }

        @JvmStatic
        fun onUnsubscribeResult(context: Context, category: String, resultCode: Long, reason: String?, topic: String?) {
            synchronized(sCallbacks) {
                for (callback in sCallbacks) {
                    if (isCategoryMatch(category, callback.category)) {
                        callback.onUnsubscribeResult(resultCode, reason, topic)
                    }
                }
            }
        }

        @JvmStatic
        fun processMessageForCallback(context: Context, pushMessageInterface: PushMessageInterface) {
            if (pushMessageInterface is MiPushMessage) {
                onReceiveMessage(context, pushMessageInterface)
                return
            }
            if (pushMessageInterface is MiPushCommandMessage) {
                val command = pushMessageInterface.command
                if (Command.COMMAND_REGISTER.value == command) {
                    val commandArguments = pushMessageInterface.commandArguments
                    val str = if (commandArguments != null && commandArguments.isNotEmpty()) commandArguments[0] else null
                    onInitializeResult(pushMessageInterface.resultCode, pushMessageInterface.reason, str)
                    return
                }
                if (Command.COMMAND_SET_ALIAS.value == command || Command.COMMAND_UNSET_ALIAS.value == command || Command.COMMAND_SET_ACCEPT_TIME.value == command) {
                    onCommandResult(
                        context,
                        pushMessageInterface.category ?: "",
                        command,
                        pushMessageInterface.resultCode,
                        pushMessageInterface.reason,
                        pushMessageInterface.commandArguments
                    )
                    return
                }
                if (Command.COMMAND_SUBSCRIBE_TOPIC.value == command) {
                    val commandArguments = pushMessageInterface.commandArguments
                    onSubscribeResult(
                        context,
                        pushMessageInterface.category ?: "",
                        pushMessageInterface.resultCode,
                        pushMessageInterface.reason,
                        if (commandArguments == null || commandArguments.isEmpty()) null else commandArguments[0]
                    )
                } else if (Command.COMMAND_UNSUBSCRIBE_TOPIC.value == command) {
                    val commandArguments = pushMessageInterface.commandArguments
                    onUnsubscribeResult(
                        context,
                        pushMessageInterface.category ?: "",
                        pushMessageInterface.resultCode,
                        pushMessageInterface.reason,
                        if (commandArguments == null || commandArguments.isEmpty()) null else commandArguments[0]
                    )
                }
            }
        }

        @JvmStatic
        fun removeAllPushCallbackClass() {
            synchronized(sCallbacks) {
                sCallbacks.clear()
            }
        }

        @JvmStatic
        fun removeAllUPSCallback() {
            synchronized(sICallbackResult) {
                sICallbackResult.clear()
            }
        }

        @JvmStatic
        fun removePushCallbackClass(miPushClientCallback: MiPushClient.MiPushClientCallback) {
            synchronized(sCallbacks) {
                sCallbacks.remove(miPushClientCallback)
            }
        }

        @JvmStatic
        fun removeUPSCallback(iCallbackResult: MiPushClient.ICallbackResult<*>) {
            synchronized(sICallbackResult) {
                sICallbackResult.remove(iCallbackResult)
            }
        }

        private fun scheduleJob(context: Context, intent: Intent) {
            sPool.execute {
                onHandleIntent(context, intent)
            }
        }

        @JvmStatic
        fun startService(context: Context) {
            val intent = Intent().apply {
                component = ComponentName(context, PushMessageHandler::class.java)
            }
            try {
                context.startService(intent)
            } catch (e: Exception) {
                MyLog.w(e.message ?: "")
            }
        }
    }

    interface PushMessageInterface : java.io.Serializable

    override fun hasJob(): Boolean {
        return sPool.queue?.size ?: 0 > 0
    }

    override fun onBind(intent: Intent): IBinder? = null

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        val result = super.onStartCommand(intent, flags, startId)
        if (intent != null) {
            scheduleJob(applicationContext, intent)
        }
        return result
    }
}
